// ngx-videogular/esm5/lib/streaming/vg-streaming.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/streaming/vg-streaming.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VgDASH } from './vg-dash/vg-dash';
import { VgHLS } from './vg-hls/vg-hls';
/**
 * @record
 */
export function IDRMLicenseServer() { }
export { VgDASH } from './vg-dash/vg-dash';
export { VgHLS } from './vg-hls/vg-hls';
var VgStreamingModule = /** @class */ (function () {
    function VgStreamingModule() {
    }
    VgStreamingModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    declarations: [
                        VgDASH, VgHLS
                    ],
                    exports: [
                        VgDASH, VgHLS
                    ]
                },] }
    ];
    return VgStreamingModule;
}());
export { VgStreamingModule };
