// ngx-videogular/esm2015/lib/core/vg-media/vg-media-element.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/core/vg-media/vg-media-element.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class VgMediaElement {
    /**
     * @return {?}
     */
    get audioTracks() { return null; }
    // @ts-ignore
    /**
     * @param {?} kind
     * @param {?=} label
     * @param {?=} language
     * @return {?}
     */
    addTextTrack(kind, label, language) {
        return null;
    }
    // @ts-ignore
    /**
     * @param {?} type
     * @return {?}
     */
    canPlayType(type) {
        return null;
    }
    /**
     * @return {?}
     */
    load() {
    }
    /**
     * @return {?}
     */
    msClearEffects() {
    }
    /**
     * @return {?}
     */
    msGetAsCastingSource() {
        return null;
    }
    // @ts-ignore
    /**
     * @param {?} activatableClassId
     * @param {?} effectRequired
     * @param {?=} config
     * @return {?}
     */
    msInsertAudioEffect(activatableClassId, effectRequired, config) {
    }
    // @ts-ignore
    /**
     * @param {?} mediaKeys
     * @return {?}
     */
    msSetMediaKeys(mediaKeys) {
    }
    // @ts-ignore
    /**
     * @param {?=} mediaProtectionManager
     * @return {?}
     */
    msSetMediaProtectionManager(mediaProtectionManager) {
    }
    /**
     * @return {?}
     */
    pause() {
    }
    /**
     * @return {?}
     */
    play() {
        return null;
    }
    // @ts-ignore
    /**
     * @param {?} mediaKeys
     * @return {?}
     */
    setMediaKeys(mediaKeys) {
        return null;
    }
    // @ts-ignore
    /**
     * @param {?} type
     * @param {?} listener
     * @param {?=} useCapture
     * @return {?}
     */
    addEventListener(type, listener, useCapture) {
    }
}
if (false) {
    /** @type {?} */
    VgMediaElement.prototype.id;
    /** @type {?} */
    VgMediaElement.prototype.elem;
    /** @type {?} */
    VgMediaElement.prototype.autoplay;
    /** @type {?} */
    VgMediaElement.prototype.buffered;
    /** @type {?} */
    VgMediaElement.prototype.controls;
    /** @type {?} */
    VgMediaElement.prototype.crossOrigin;
    /** @type {?} */
    VgMediaElement.prototype.currentSrc;
    /** @type {?} */
    VgMediaElement.prototype.currentTime;
    /** @type {?} */
    VgMediaElement.prototype.defaultMuted;
    /** @type {?} */
    VgMediaElement.prototype.defaultPlaybackRate;
    /** @type {?} */
    VgMediaElement.prototype.duration;
    /** @type {?} */
    VgMediaElement.prototype.ended;
    /** @type {?} */
    VgMediaElement.prototype.error;
    /** @type {?} */
    VgMediaElement.prototype.loop;
    /** @type {?} */
    VgMediaElement.prototype.mediaKeys;
    /** @type {?} */
    VgMediaElement.prototype.msAudioCategory;
    /** @type {?} */
    VgMediaElement.prototype.msAudioDeviceType;
    /** @type {?} */
    VgMediaElement.prototype.msGraphicsTrustStatus;
    /** @type {?} */
    VgMediaElement.prototype.msKeys;
    /** @type {?} */
    VgMediaElement.prototype.msPlayToDisabled;
    /** @type {?} */
    VgMediaElement.prototype.msPlayToPreferredSourceUri;
    /** @type {?} */
    VgMediaElement.prototype.msPlayToPrimary;
    /** @type {?} */
    VgMediaElement.prototype.msPlayToSource;
    /** @type {?} */
    VgMediaElement.prototype.msRealTime;
    /** @type {?} */
    VgMediaElement.prototype.muted;
    /** @type {?} */
    VgMediaElement.prototype.networkState;
    /** @type {?} */
    VgMediaElement.prototype.onencrypted;
    /** @type {?} */
    VgMediaElement.prototype.onmsneedkey;
    /** @type {?} */
    VgMediaElement.prototype.paused;
    /** @type {?} */
    VgMediaElement.prototype.playbackRate;
    /** @type {?} */
    VgMediaElement.prototype.played;
    /** @type {?} */
    VgMediaElement.prototype.preload;
    /** @type {?} */
    VgMediaElement.prototype.readyState;
    /** @type {?} */
    VgMediaElement.prototype.seekable;
    /** @type {?} */
    VgMediaElement.prototype.seeking;
    /** @type {?} */
    VgMediaElement.prototype.src;
    /** @type {?} */
    VgMediaElement.prototype.srcObject;
    /** @type {?} */
    VgMediaElement.prototype.textTracks;
    /** @type {?} */
    VgMediaElement.prototype.videoTracks;
    /** @type {?} */
    VgMediaElement.prototype.volume;
    /** @type {?} */
    VgMediaElement.prototype.HAVE_CURRENT_DATA;
    /** @type {?} */
    VgMediaElement.prototype.HAVE_ENOUGH_DATA;
    /** @type {?} */
    VgMediaElement.prototype.HAVE_FUTURE_DATA;
    /** @type {?} */
    VgMediaElement.prototype.HAVE_METADATA;
    /** @type {?} */
    VgMediaElement.prototype.HAVE_NOTHING;
    /** @type {?} */
    VgMediaElement.prototype.NETWORK_EMPTY;
    /** @type {?} */
    VgMediaElement.prototype.NETWORK_IDLE;
    /** @type {?} */
    VgMediaElement.prototype.NETWORK_LOADING;
    /** @type {?} */
    VgMediaElement.prototype.NETWORK_NO_SOURCE;
}
