// ngx-videogular/esm5/lib/core/vg-media/vg-media-element.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/core/vg-media/vg-media-element.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var VgMediaElement = /** @class */ (function () {
    function VgMediaElement() {
    }
    Object.defineProperty(VgMediaElement.prototype, "audioTracks", {
        get: /**
         * @return {?}
         */
        function () { return null; },
        enumerable: true,
        configurable: true
    });
    // @ts-ignore
    // @ts-ignore
    /**
     * @param {?} kind
     * @param {?=} label
     * @param {?=} language
     * @return {?}
     */
    VgMediaElement.prototype.addTextTrack = 
    // @ts-ignore
    /**
     * @param {?} kind
     * @param {?=} label
     * @param {?=} language
     * @return {?}
     */
    function (kind, label, language) {
        return null;
    };
    // @ts-ignore
    // @ts-ignore
    /**
     * @param {?} type
     * @return {?}
     */
    VgMediaElement.prototype.canPlayType = 
    // @ts-ignore
    /**
     * @param {?} type
     * @return {?}
     */
    function (type) {
        return null;
    };
    /**
     * @return {?}
     */
    VgMediaElement.prototype.load = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    VgMediaElement.prototype.msClearEffects = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    VgMediaElement.prototype.msGetAsCastingSource = /**
     * @return {?}
     */
    function () {
        return null;
    };
    // @ts-ignore
    // @ts-ignore
    /**
     * @param {?} activatableClassId
     * @param {?} effectRequired
     * @param {?=} config
     * @return {?}
     */
    VgMediaElement.prototype.msInsertAudioEffect = 
    // @ts-ignore
    /**
     * @param {?} activatableClassId
     * @param {?} effectRequired
     * @param {?=} config
     * @return {?}
     */
    function (activatableClassId, effectRequired, config) {
    };
    // @ts-ignore
    // @ts-ignore
    /**
     * @param {?} mediaKeys
     * @return {?}
     */
    VgMediaElement.prototype.msSetMediaKeys = 
    // @ts-ignore
    /**
     * @param {?} mediaKeys
     * @return {?}
     */
    function (mediaKeys) {
    };
    // @ts-ignore
    // @ts-ignore
    /**
     * @param {?=} mediaProtectionManager
     * @return {?}
     */
    VgMediaElement.prototype.msSetMediaProtectionManager = 
    // @ts-ignore
    /**
     * @param {?=} mediaProtectionManager
     * @return {?}
     */
    function (mediaProtectionManager) {
    };
    /**
     * @return {?}
     */
    VgMediaElement.prototype.pause = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    VgMediaElement.prototype.play = /**
     * @return {?}
     */
    function () {
        return null;
    };
    // @ts-ignore
    // @ts-ignore
    /**
     * @param {?} mediaKeys
     * @return {?}
     */
    VgMediaElement.prototype.setMediaKeys = 
    // @ts-ignore
    /**
     * @param {?} mediaKeys
     * @return {?}
     */
    function (mediaKeys) {
        return null;
    };
    // @ts-ignore
    // @ts-ignore
    /**
     * @param {?} type
     * @param {?} listener
     * @param {?=} useCapture
     * @return {?}
     */
    VgMediaElement.prototype.addEventListener = 
    // @ts-ignore
    /**
     * @param {?} type
     * @param {?} listener
     * @param {?=} useCapture
     * @return {?}
     */
    function (type, listener, useCapture) {
    };
    return VgMediaElement;
}());
export { VgMediaElement };
if (false) {
    /** @type {?} */
    VgMediaElement.prototype.id;
    /** @type {?} */
    VgMediaElement.prototype.elem;
    /** @type {?} */
    VgMediaElement.prototype.autoplay;
    /** @type {?} */
    VgMediaElement.prototype.buffered;
    /** @type {?} */
    VgMediaElement.prototype.controls;
    /** @type {?} */
    VgMediaElement.prototype.crossOrigin;
    /** @type {?} */
    VgMediaElement.prototype.currentSrc;
    /** @type {?} */
    VgMediaElement.prototype.currentTime;
    /** @type {?} */
    VgMediaElement.prototype.defaultMuted;
    /** @type {?} */
    VgMediaElement.prototype.defaultPlaybackRate;
    /** @type {?} */
    VgMediaElement.prototype.duration;
    /** @type {?} */
    VgMediaElement.prototype.ended;
    /** @type {?} */
    VgMediaElement.prototype.error;
    /** @type {?} */
    VgMediaElement.prototype.loop;
    /** @type {?} */
    VgMediaElement.prototype.mediaKeys;
    /** @type {?} */
    VgMediaElement.prototype.msAudioCategory;
    /** @type {?} */
    VgMediaElement.prototype.msAudioDeviceType;
    /** @type {?} */
    VgMediaElement.prototype.msGraphicsTrustStatus;
    /** @type {?} */
    VgMediaElement.prototype.msKeys;
    /** @type {?} */
    VgMediaElement.prototype.msPlayToDisabled;
    /** @type {?} */
    VgMediaElement.prototype.msPlayToPreferredSourceUri;
    /** @type {?} */
    VgMediaElement.prototype.msPlayToPrimary;
    /** @type {?} */
    VgMediaElement.prototype.msPlayToSource;
    /** @type {?} */
    VgMediaElement.prototype.msRealTime;
    /** @type {?} */
    VgMediaElement.prototype.muted;
    /** @type {?} */
    VgMediaElement.prototype.networkState;
    /** @type {?} */
    VgMediaElement.prototype.onencrypted;
    /** @type {?} */
    VgMediaElement.prototype.onmsneedkey;
    /** @type {?} */
    VgMediaElement.prototype.paused;
    /** @type {?} */
    VgMediaElement.prototype.playbackRate;
    /** @type {?} */
    VgMediaElement.prototype.played;
    /** @type {?} */
    VgMediaElement.prototype.preload;
    /** @type {?} */
    VgMediaElement.prototype.readyState;
    /** @type {?} */
    VgMediaElement.prototype.seekable;
    /** @type {?} */
    VgMediaElement.prototype.seeking;
    /** @type {?} */
    VgMediaElement.prototype.src;
    /** @type {?} */
    VgMediaElement.prototype.srcObject;
    /** @type {?} */
    VgMediaElement.prototype.textTracks;
    /** @type {?} */
    VgMediaElement.prototype.videoTracks;
    /** @type {?} */
    VgMediaElement.prototype.volume;
    /** @type {?} */
    VgMediaElement.prototype.HAVE_CURRENT_DATA;
    /** @type {?} */
    VgMediaElement.prototype.HAVE_ENOUGH_DATA;
    /** @type {?} */
    VgMediaElement.prototype.HAVE_FUTURE_DATA;
    /** @type {?} */
    VgMediaElement.prototype.HAVE_METADATA;
    /** @type {?} */
    VgMediaElement.prototype.HAVE_NOTHING;
    /** @type {?} */
    VgMediaElement.prototype.NETWORK_EMPTY;
    /** @type {?} */
    VgMediaElement.prototype.NETWORK_IDLE;
    /** @type {?} */
    VgMediaElement.prototype.NETWORK_LOADING;
    /** @type {?} */
    VgMediaElement.prototype.NETWORK_NO_SOURCE;
}
