// ngx-videogular/esm2015/lib/core/vg-core.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/core/vg-core.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { VgControlsHidden } from './services/vg-controls-hidden';
import { VgFullscreenAPI } from './services/vg-fullscreen-api';
import { VgUtils } from './services/vg-utils';
import { VgAPI } from './services/vg-api';
import { VgMedia } from './vg-media/vg-media';
import { VgCuePoints } from './vg-cue-points/vg-cue-points';
import { VgPlayer } from './vg-player/vg-player';
import { VgStates } from './states/vg-states';
import { VgEvents } from './events/vg-events';
export {} from './vg-media/i-media-element';
export {} from './vg-media/i-playable';
export { VgMediaElement } from './vg-media/vg-media-element';
export { VgAPI } from './services/vg-api';
export { VgFullscreenAPI } from './services/vg-fullscreen-api';
export { VgUtils } from './services/vg-utils';
export { VgControlsHidden } from './services/vg-controls-hidden';
export { VgEvents } from './events/vg-events';
export { VgStates } from './states/vg-states';
/**
 * @record
 */
export function BitrateOption() { }
if (false) {
    /** @type {?} */
    BitrateOption.prototype.qualityIndex;
    /** @type {?} */
    BitrateOption.prototype.width;
    /** @type {?} */
    BitrateOption.prototype.height;
    /** @type {?} */
    BitrateOption.prototype.bitrate;
    /** @type {?} */
    BitrateOption.prototype.mediaType;
    /** @type {?|undefined} */
    BitrateOption.prototype.label;
}
export class VgCoreModule {
}
VgCoreModule.decorators = [
    { type: NgModule, args: [{
                declarations: [VgMedia, VgCuePoints, VgPlayer],
                exports: [VgMedia, VgCuePoints, VgPlayer],
                providers: [VgAPI, VgFullscreenAPI, VgUtils, VgControlsHidden, VgStates, VgEvents]
            },] }
];
